### DISCUSSION TITLE
Acknowledgment lora

### DISCUSSION THREAD

**Discussion by julianwidyap** [sandeepmistry/arduino-LoRa#505] [Q&A]
Hello i will ask about ack mechanism in lora, i have a project and than i must used ack mechanism. Can lora doing acknowledgment mechanism, and how to alghoritm program for that thank you.

**Comment by IoTThinks** [comment]
LoRa is raw communication. Except sending and receiving, you have to do everything including ack for your own.
The sender sends the message with message id and the receiver sends back a dummy ack with message id.

LoRaWAN protocol has built-in ACK mechanism, but the protocol is much more complex.

**Reply by halukmy** [reply]
can you sample ack usage with lora?

**Comment by Kongduino** [comment]
As IoTThinks says LoRa P2P is raw data, so if you want to implement ACK, you have to send a UUID with your message, and the receiver sends an empty ACK/PONG message with that UUID. After the sender has sent the message, you store the ID with a `false` flag, and wait until you receive the proper ACK message to set it to `true`.

Once you got that working, you can make it evolve to provide more useful functionalities: timeout, resend yes/no, how many resends before fail, etc.
